Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE", as well as your collaborators below:

In [11]:
COLLABORATORS = ""

---

In [12]:
import numpy as np

<div class="alert alert-info">**Hint**: Much of the material covered in this problem is introduced in the AIMA3 pp. 802-813 or AIMA2 pp. 712-722 readings. If you are having trouble with the questions in this notebook, this might be a good place to look.</div>

 One day, you find yourself standing in a musty room in the back of an old magic shop. Heavy velvet curtains cover the dirty windows, reluctantly letting in a few narrow beams of light, which succeed only in illuminating layers of dust suspended in the stale air. You glance around cautiously, trying not to invite scrutiny from the watchful shopkeep. Your eyes dance over row upon row of leather-bound books, resting on the rich mahogany shelves lining each wall. Who knows how long they've been here, or what ancient secrets they conceal. They almost call out to you to be opened, to divulge their long-forgotten knowledge. But no... you've come here for another reason.

  Before you can even turn around, the shopkeep anticipates your question. "You've come for a coin, haven't you?" His voice sounds strangely distant, and reminds you of some kind of large, creepy bird or something. You nod, swallowing. He totters past you, one of his legs struggling to keep up with the other. Reaching a black armoire in the corner of the room — how did you not notice it before? — he stops. The shopkeep slowly opens one of the drawers, revealing a beautiful, glimmering coin. It looks perfectly untarnished, yet, somehow, emanates the energy of centuries past. 

  Your father's deep, soulful voice echoes in your head. "Fetch me a coin of Azeroth, child, but only if its probability of landing heads is between 0.55 and 0.75." 
  
  <hr/>
  
  Is the coin in the armoire such a coin? How can we determine if this coin matches father's request?

  We can formalize your predicament by letting the variable $\theta$ denote a coin of Azeroth's probability of landing heads on each toss, which is what you need to infer. We assume that each toss is independent of the others. In this learning problem, your hypothesis space is the set of all possible values of $\theta$, which is all the real numbers from 0 to 1. We need to compare an infinite number of hypotheses! How can we do this?

In parts A through C, you will write functions that estimate the weight of the coin. Each function takes a single argument, a sequence of tosses, a $1 \times n$ binary row vector representing an observed sequence of coin tosses where `1`'s represent heads and `0`'s represent tails.

## Part A (1 point)

<div class="alert alert-success">Using the standard frequentist method of *maximum likelihood estimation*, define a function that, given a sequence of tosses, returns an estimate for the value of $\theta$, the probability that the coin lands heads up. Remember that *MLE* relies only on data that we've actually observed, so the prior should NOT factor into your answer. The equation for MLE can be found in the lecture notes.</div>

In [13]:
def mle_azeroth(sequence):
    """Uses MLE to estimate the value of theta.
    
    Hint: Your solution can be done in 1 line of code.
    
    Parameters
    ----------
    sequence : an (m,) Numpy array of 1s and 0s.
        The observed sequence of coin flips.
        
    Returns
    -------
    The value of theta, a float.
    
    """
    # YOUR CODE HERE
    return (np.sum([s for s in sequence if s ==1]) / len(sequence))

In [14]:
"""Check that the function is correct."""
from nose.tools import assert_equal

assert(mle_azeroth)

sequence = np.array([1, 1, 0, 1, 0, 1, 1, 1, 1, 1])
assert(mle_azeroth(sequence) == 0.8)

sequence = np.array([1, 1, 1, 1, 1])
assert(mle_azeroth(sequence) == 1.0)

sequence = np.array([0, 0, 0, 0, 0])
assert(mle_azeroth(sequence) == 0.0)

sequence = np.array([1, 0])
assert(mle_azeroth(sequence) == 0.5)

sequence = np.array([1, 0, 1, 0])
assert(mle_azeroth(sequence) == 0.5)

sequence = np.array([1, 0, 1, 0, 1, 0])
assert(mle_azeroth(sequence) == 0.5)

print("Success!")

Success!


## Part B (1 point)

<div class="alert alert-success">Assuming a Bernoulli likelihood and Beta prior, define a function that computes the *maximum a posteriori* (MAP) estimate for $\theta$. Your prior should be defined by two variables, `prior_heads` and `prior_tails`, which act as pseudocounts of the number of previously seen heads and tails.) The equation for computing the MAP estimate can be found in the lecture notes.</div>

In [15]:
def map_azeroth(sequence, prior_heads=0, prior_tails=0):
    """Computes the MAP estimate of theta.
    
    Hint: Your solution can be done in a line or two of code.
    
    Parameters
    ----------
    sequence : an (m,) Numpy array of 1s and 0s.
        The observed sequence of coin flips.
    prior_heads : an integer psedocount representing the
        prior beliefs that the coin is biased towards heads.
    prior_tails : an integer psedocount representing the
        prior beliefs that the coin is biased towards tails.
        
    Returns
    -------
    The value of theta, a float.
    
    """
    # YOUR CODE HERE
    return ((np.sum([s for s in sequence if s ==1])) + prior_heads) / (len(sequence) + prior_tails + prior_heads)

In [16]:
"""Check that the function exists."""
assert(map_azeroth)

"""Check that, with no prior specified, the MAP and MLE estimates are the same."""
for i in range(10):
    sequence = np.random.randint(0, 1, size=(10,))
    assert(mle_azeroth(sequence) == map_azeroth(sequence))

"""Check that changing the prior changes the estimate."""
sequence = np.array([1, 1, 0, 1, 0, 1, 1, 1, 1, 1])
assert(map_azeroth(sequence) == 0.8)
assert(map_azeroth(sequence, 1, 1) == 0.75)
assert(map_azeroth(sequence, 5, 5) == 0.65)
assert(map_azeroth(sequence, 1, 9) == 0.45)
assert(map_azeroth(sequence, 15, 205) == 0.1)
assert(map_azeroth(np.array([]), 1, 1) == 0.5)
assert(map_azeroth(np.array([1, 0]), 1, 1) == 0.5)
assert(map_azeroth(np.array([1, 1, 1, 1, 1, 0]), 1, 1) == 0.75)
assert(map_azeroth(np.array([0, 0, 0, 0, 0, 1]), 1, 1) == 0.25)

print("Success!")

Success!


## Part C (1 point)

<div class="alert alert-success">Assuming a Bernoulli likelihood and Beta prior, define a function that computes the *posterior mean* estimate for $\theta$. Your prior should be defined by two variables, `prior_heads` and `prior_tails`, which act as pseudocounts of the number of previously seen heads and tails.)</div>

In [17]:
def pm_azeroth(sequence, prior_heads=0, prior_tails=0):
    """Computes the posterior mean estimate of theta.
    
    Hint: Your solution can be done in a line or two of code.
    
    Parameters
    ----------
    sequence : an (m,) Numpy array of 1s and 0s.
        The observed sequence of coin flips.
    prior_heads : an integer psedocount representing the
        prior beliefs that the coin is biased towards heads.
    prior_tails : an integer psedocount representing the
        prior beliefs that the coin is biased towards tails.
        
    Returns
    -------
    The value of theta, a float.
    
    """
    # YOUR CODE HERE
    return ((np.sum([s for s in sequence if s ==1])) + prior_heads + 1) / (len(sequence) + prior_tails + prior_heads + 2)

In [18]:
"""Check that the function exists."""
assert(pm_azeroth)


"""Check that changing the prior changes the estimate."""
sequence = np.array([1, 1, 0, 1, 0, 1, 1, 1, 1, 1])
assert(pm_azeroth(sequence) == 0.75)
assert(pm_azeroth(sequence, 1, 7) == 0.5)
assert(pm_azeroth(sequence, 7, 1) == 0.8)
assert(pm_azeroth(sequence, 27, 1) == 0.9)
assert(pm_azeroth(sequence, 124, 130) == 0.5)
assert(pm_azeroth(np.array([]), 1, 1) == 0.5)
assert(pm_azeroth(np.array([1, 0]), 1, 1) == 0.5)
assert(pm_azeroth(np.array([1, 1, 1, 1, 1, 0]), 1, 1) == 0.70)
assert(pm_azeroth(np.array([0, 0, 0, 0, 0, 1]), 1, 1) == 0.30)

print("Success!")

Success!


## Part D (1 point)

Assume you tossed a coin of Azeroth 10 times and observed the specific sequence `HHTHTHHHHH`. Further assume that we've observed some fictitious trials and have a prior belief that coins of Azeroth are slightly biased towards landing heads. Thus, let $V_H$ = 55 and $V_T$ = 45, denoting that in the past we've observed the coin of Azeroth flipped 100 times and saw 55 heads and 45 tails. Let's see the results of our estimators:

In [19]:
sequence = np.array([1, 1, 0, 1, 0, 1, 1, 1, 1, 1])
print("MLE: {}".format(mle_azeroth(sequence)))
print("MAP: {}".format(map_azeroth(sequence, prior_heads=55, prior_tails=45)))
print("Posterior mean: {}".format(pm_azeroth(sequence, prior_heads=55, prior_tails=45)))

MLE: 0.8
MAP: 0.5727272727272728
Posterior mean: 0.5714285714285714


<div class="alert alert-success">Given the father's request for a coin with probability of landing heads between $0.55$ and $0.75$ - do you think you should you take this coin? Why or why not?</div>

Yes, I believe I should take this coin because the MLE value results in 0.8, and the MAP value results in 0.5727. Since it is known that the Azeroth coin was flipped 100 times, and saw 55 heads and 45 tails previously, the MAP value takes this into account and therefore the result of 0.5727 seems like a pretty accurate probability of landing heads between 0.55 and 0.75. In addition, the fact that the MLE value is 0.8 is a high upper bound gives a strong certainty that the coin is more biased towards heads. Finally, since the posterior mean is right above 0.55 (0.571), we can be farily certain that it has a probability of landing heads between 0.55 and 0.75 -- so we should take this coin!

---

Before turning this problem in remember to do the following steps:

1. **Restart the kernel** (Kernel$\rightarrow$Restart)
2. **Run all cells** (Cell$\rightarrow$Run All)
3. **Save** (File$\rightarrow$Save and Checkpoint)

<div class="alert alert-danger">After you have completed these three steps, ensure that the following cell has printed "No errors". If it has <b>not</b> printed "No errors", then your code has a bug in it and has thrown an error! Make sure you fix this error before turning in your problem set.</div>

In [20]:
print("No errors!")

No errors!
